2024-10-30-lapply/cc_lapply.Rmd

```diff
@@ -0,0 +1,694 @@
+  pull()
+
+# Check practice codes
+head(practice_codes_vec,50)
+
+# Check length
+length(practice_codes_vec)
+```
+
+We can now apply the function to our vector.
+We will time how long it takes using 'tic()' and 'toc()' from the 'tictoc' package.
+If a function has more than input:
+
++ The 'X' parameter is your vector
++ Explicitly state the other parameter inputs
+
+The code is hashed out, but would take ~3.5 minutes to run.
+
+```{r}
+# # Start timer
+# tic()
+# 
+# # Lapply
+# results = lapply(
+#   X = practice_codes_vec,
+#   FUN = sum_items_by_practice,
+#   df = df
+#   )
+# 
+# # End timer
+# toc()
+```
+
+### Run in parLapply
+
+This is a fairly slow process, that could be easily sped up.
+We can use parLapply to run the process in parallel.
+This is a very similar code structure to 'regular' lapply. 
+
+There are a few considerations to bear in mind when using parLapply:
+
++ How many cores your machine has
++ How many cores you want to *not* be used
++ The size of your function inputs/dependencies
++ E.g., if you want x6 processing, and you have a 500MB dataframe, this needs copying x6 times.
+
+The process then has a few steps:
+
+1. Define how many cores you want to use
+2. 'Make' your cluster
+3. Export all the required libraries (used in the function) to each cluster
+4. Export all the required objects (used in the function) to each cluster
+5. Generate the results
+6. Stop the cluster (IMPORTANT!)
+
+The function itself:
+
++ You need to specify the cluster object, in the 'cl' parameter.
++ The 'FUN' parameter becomes 'fun'
+
+Can first choose number of cores and setup cluster.
+
+```{r}
+# Generate appropriate number of cores
+n_cores = detectCores() - 2
+
+# Set up parallel
+clust = makeCluster(n_cores)
+```
+
+The function only requires the 'dplyr' library.
+And it only requires our vector of practice codes, and we can specify where to find these (the global environment).
+
+```{r}
+# Export libraries to cluster
+clusterEvalQ(
+  cl = clust,
+  {
+    library(dplyr); 
+  }
+)
+
+# Export required objects to cluster
+clusterExport(
+  cl = clust,
+  varlist = c(
+    "practice_codes_vec"
+  ),
+  envir = environment()
+)
+```
+
+We can now run the code in parallel.
+The longer the runtime, the 'better' the time saving will be.
+This is because some overhead is required in the process.
+These include:
+
++ Forking
++ Initialization
++ Data Serialisation
++ Memory allocation
++ and so on ...
+
+That said, we can still use this to speed up a process that in the first instance took several minutes.
+
+```{r}
+# Start timer
+tic()
+
+# Generate cqc details: 150 seconds
+results = parallel::parLapply(
+  cl = clust, 
+  X = practice_codes_vec, 
+  fun = sum_items_by_practice,
+  df = df
+)
+
+# End timer
+toc()
+```
+
+This took ~1 minute compared to ~3.5 minutes.
+This is not a x6 saving, due to the aforementioned overhead.
+Final thing is to stop the cluster.
+
+```{r}
+# Stop Cluster
+parallel::stopCluster(clust)
```

Comment:
Could namespace all calls to `parallel` package functions, as it would make clear which functions these are (and not just base R).